### ISSUE TITLE
Integrate patsy?

### ISSUE THREAD

**Issue by twiecki** [pymc-devs/pymc3#199]
Patsy is a linear model specification package. It's currenty mainly used by statsmodels: http://jseabold.net/presentations/seabold_pydata2012.html#slide1

Here is an example using it for a simple linear regression with pymc3:
https://github.com/twiecki/pymc/blob/patsy/examples/regression_patsy.py

It's very powerful and allows flexible creation of complex linear models. Might be a nice addition to pymc3.

**Comment by jsalvatier** [comment]
That does seem interesting. How would you _want_ to use this in PyMC? 

I'm thinking you would want to be able to generate a linear term easily. Something like.

```
y_est, coeff = linear_component(model, "y ~ a + a*b", data)
```

Which would generate a random variable of coefficients and an estimate of y (which depends on coeffs). You would need a source of priors for the coeffs though. 

Above, I'm imagining that data could be a DataFrame of data or some kind of theano variables, but I don't know feasible the later is since I don't think pandas and Theano mix that well.

You would also probably want to do: 

```
coeff = linear_model(model, "y ~ a + a *b", y, data, error = Normal, prior = ...)
```

Perhaps we should draw inspiration from R.

**Comment by twiecki** [comment]
DataFrames are just numpy arrays underneath. Your last syntax looks potentially interesting.

**Comment by njsmith** [comment]
If your design matrices are fixed, then using patsy will be pretty trivial -- just a bit of pretty API on top of the code that @twiecki linked at the top. (Let me know if you run into limitations, of course.)

Would be even more awesome would be if one could make it work even when design matrices aren't fixed. Like if you write "y ~ x1 + x2" and x1 and x2 are latent rather than observed. I think this is... intriguingly close to possible, but we might have to sit down and think about it a bit. You can think of patsy as a system that, given an abstract formula like "y ~ x1 + x2", produces a very complicated black box which takes assignments of values to those variables and hands you back a design matrix. What else do you need? Derivatives of the design matrix WRT the latent variables? (Do you handle categorical variables?)

**Comment by njsmith** [comment]
(Oh yes, and WRT the comment about pandas versus theano above -- patsy doesn't depend on pandas at all. And it has tons of tools for letting you peek inside the black box in systematic ways.)

**Comment by jsalvatier** [comment]
Since patsy.dmatrix already accepts a dict of variables, perhaps all we need to do is build that dict intelligently? I guess probably not because, we need to replace any numpy operations with theano operations. Patsy probably uses a concatenate internally, which will try to do a numpy concatenate not a theano concatenate. However, that seems close to working.

**Comment by njsmith** [comment]
Conceptually there are a few steps in going from a patsy formula to a
design matrix. First, each factor might have some arbitrary Python code
that gets run to get the actual values (e.g. if you have 'y ~ np.log(x1 +
x2)', then the 'np.log(x1 + x2)' part just gets passed to eval()). I
wouldn't suggest messing around trying to replace numpy operations with
theano operations in this stage -- instead I'd just tell people to use
'T.log' instead of 'np.log' :-).

Then, we look at the values we got for all of the factors, and we work out
by inspection: (1) whether they're numerical or categorical, (2) if they're
numeric, how many columns they have, (3) if they're categorical, how many
levels they have, what those levels are called, and which contrast coding
scheme we should be using. This information is enough to work out the
overall shape, if you will, of the final design matrix: at this point we
know exactly what the columns in the output will be and how to compute
them, even if we don't know what data will go into them. This information
gets compiled into a 'DesignMatrixBuilder'. It's a bit more than a
concatenate :-) But if you have the relevant information available
somewhere, we can certainly find some way to accomplish this part.

Finally, given a 'DesignMatrixBuilder' and some data values, you can get
out a design matrix. This part is pretty simple, and might already be fine
for you? Do you need it to be a theano op instead of a python function, for
GPU computation or something? Do you need the gradient? Do you support
categorical variables? (Those are by far the most complex part of patsy's
builder code, and you can't compute gradients for them, obviously.)

**Comment by twiecki** [comment]
OK, I made some progress on this. Currently all code is stand-alone on the glm branch (based on scaling branch). All functionality currently rests in an IPython NB (http://nbviewer.ipython.org/urls/raw.github.com/pymc-devs/pymc/glm/pymc/examples/GLM.ipynb).

It depends on `statsmodels` for initializing (which helps a lot) and `patsy` for building the design-matrix. Not sure if that is too dependency heavy. Could be optional I suppose.

The IPyNB shows the core ideas of what is possible. Having played with this for a little while now it's actually not always easy to get it set up correctly when doing this manually and the `glm` function helps with that quite a bit.

@njsmith I currently do this: `dmatrix = patsy.dmatrix(glm_descr.split('~')[1], data)` which seems wrong. There certainly is a better way to split rhs and lhs but I don't know how.

I haven't experimented with e.g.  'y ~ T.log(x1 + x2)' as you suggested but it would be cool if this was working.

**Comment by jsalvatier** [comment]
Looking good~ In the Hierarchial example, a better way to pass in `intercept` and `slope` is construct a distribution using `Normal.dist(sat_data.sat_t.mean(), sat_data.sat_t.std())` and similarly for slope. I think this should be easier to read.

How does this compare to the R version of GLM?

**Comment by jsalvatier** [comment]
Seems like `glm` could take kwargs for x,y. Also, defaults seem like they would be nice for sigma and the observed distribution.

Do you want to do a pull request?

**Comment by twiecki** [comment]
Yeah, I first wanted to get some feedback on the interface etc. So you are
suggesting to put the whole model construction in that function by passing
in the likelihood. Yeah, that's a good idea.

I wonder if some regular models should already be included (like
LogisticRegression, RobustRegression etc). Behind the scenes those would
call the multipurpose function.

**Comment by jsalvatier** [comment]
It could be good to have both functions a `glm_predictor` (like the current function) and a `glm` function.

**Comment by njsmith** [comment]
Yeah, the `.split("~")` thing is definitely wrong. Consider a model where our data records failures, but we decide it'd be more interpretable to model using successes by inverting this: `"y ~ I(~failed)"`.

I don't quite understand what you're trying to do with the interface, but: if you don't use the left-hand side (I don't think you do?), then you just define the API so that people only specify the right hand side, no `~`. If you do want the left-hand side, and people will be specifying data for it and everything, then the easiest thing to do would be to just call `dmatrices` and then only use one of the resulting matrices.

If you really want to make people specify a left-hand side, but not provide any values for it, so dmatrices doesn't work, then you could do

```
# assuming things like `np.log` should be looked up in your caller's namespace
# btw you should be specifying the eval_env arg to dmatrix/dmatrices too.
env = EvalEnvironment.capture(1)
desc = ModelDesc.from_formula(formula, env)
rhs_only_desc = ModelDesc([], desc.rhs_termlist)
mat = dmatrix(rhs_only_desc, ...)
```

If you want to get both `y` and `x` values out, and you also want to know the name of the thing on the LHS so you can name a pymc graph node after it, then you can do

```
desc = ModelDesc.from_formula(formula, env)
lhs_observed, rhs_design = dmatrices(desc, ...)
lhs_name = desc.lhs_termlist[0].name()
```

Patsy has no Theano awareness right now. Situation's the same as above, I think -- in principle it's probably easy, but I have no idea what, exactly, you guys would need :-).

**Comment by njsmith** [comment]
Hah, that last example was way overcomplicated. Of course the easy way to do that is

```
lhs_observed, rhs_design = dmatrices(formula, ...)
lhs_name = lhs_observed.design_info.term_names[0]
```

**Comment by twiecki** [comment]
@njsmith Thanks! Your last option works perfectly.

@jsalvatier Two issues with trying to implement your suggestions:
1. Passing in e.g. `Normal.dist()` for the Intercept prior instead of the descriptor and then calling `.makevar()` on this seems intuitively correct. However, it seems that the result random var does not get put into the model context. How can this be done?
2. Passing in the likelihood function appears non-trivial as it's not immediate which parameter should be set to `y_est`. For the normal it's `mu` (but then `sd` needs to be set), for Bernoulli it's `p`. Any idea on how this could be done in a clean way? I suppose the obvious way would be to pass in the name of the parameter but that seems a little verbose somehow.

**Comment by twiecki** [comment]
I also updated the IPyNB (which is currently broken): http://nbviewer.ipython.org/urls/raw.github.com/pymc-devs/pymc/glm/pymc/examples/GLM.ipynb

**Comment by jsalvatier** [comment]
1. Try 

```
model = modelcontext(model);
model.Var(name, dist)
```

That will add it to the model.
2. That does seem challenging. Maybe passing in the likelihood is not the right way, maybe you just specify a `family` parameter as a string or constant value like in R and have several special cases.

**Comment by twiecki** [comment]
1. OK, that works (IPyNB is updated, also added some doc string)
2. That might be the best option actually. This might also help to remedy another problem: currently one has to pass 2 link-functions, one for the tensor and one for the initialization via statsmodels. With a look-up table those two could be bundled together and one would specify e.g. `family='Binomial'` and `link='sigmoid'`. Or we could copy the statsmodels approach where there are different Family classes for the residuals which take an optional `link` argument (e.g. binomial has sigmoid as the default link).
   Since we would split this up into 2 different functions (i.e. `linear_component` and `glm`) we also wouldn't lose generality.

**Comment by twiecki** [comment]
I implemented the second idea now (see http://nbviewer.ipython.org/urls/raw.github.com/pymc-devs/pymc/glm/pymc/examples/GLM.ipynb). Not sure if the code complexity isn't too off-putting but at least it makes the following calls to construct a GLM much easier. And the manual option is always there.

If we are happy with this approach I could move the helper functions to its own file (`pymc/glm.py`?) and write some unittests and docs.

**Comment by jsalvatier** [comment]
Lookin good! I think this mostly ready for a RP if you want to do those unittests /docs and move your examples to a separate file.

Couple comments:
1. Given the implementation maybe it makes sense to make the families be
   functions that take link function?  That is the way R does it. It would
   also be easier for others to extend.
2. `y_est, coeffs = linear_component(*args, **kwargs)` looks like it
   could go out of the if statement.
3. `start = glm('y ~ x', dict(x=x, y=y))` might be easier to read than `
   start = glm('y ~ x', {'x': x, 'y': y})`.
4. You do `trace_df = pd.DataFrame({name: np.squeeze(trace.vals) for name,
   trace in trace_htwt.samples.iteritems()})` a few times, maybe that would be
   good a good pymc helper function to have.

**Comment by twiecki** [comment]
Thanks for having a look.
1. Yeah, this is also the way statsmodels does it. Maybe it's not a bad idea. This class could then also generate the sigma parameter (when the family is normal) making the whole mechanism more modular and easier to extend. I implemented it in the last commit. I'm not quite happy with it yet though but it seems like a promising direction.
2. True, that's actually what I was meaning to do.
3. Done.
4. Yeah, I think that's a nice helper function. I replaced the calls.
